### ISSUE TITLE
[python] Fix non-appendability of `SparseNDArray`

### ISSUE THREAD

**Issue by johnkerl** [single-cell-data/TileDB-SOMA#931]
## Use-case analysis

* For cell census, arrays are immutable and people want to take their shapes
* For `soma.io`, arrays are mutable and people want to take their shapes
* Currently only the first use-case is supported

## Why this affects only `SOMASparseNDArray` and not `SOMADataFrame`

* `DataFrame` is a sparse 1D, n-attr TileDB array
  * Currently this is written with full-int64 domain at the schema level
  * You could get its non-empty domain at the data level to get the 'shape' i.e. to figure out the number of rows ...
  * ... but what we actually use to get the 'shape' aka `df.count` is
    * https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/apis/python/src/tiledbsoma/dataframe.py#L94-L100
    * https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/libtiledbsoma/src/soma_reader.cc#L176
  * This does:
    * If there is one fragment it can quickly return cell count from `FragmentInfo`
    * Else if there is more than one fragment but all with non-overlapping non-empty domains it can quickly return sum of cell count over each FragmentInfo
    * Else the computation is a bit slower
    * See also: https://github.com/single-cell-data/TileDB-SOMA/issues/383
    * See also: https://github.com/single-cell-data/TileDB-SOMA/pull/404
* `SparseNDarray` is a sparse 2D, 1-attr TileDB array
  * Currently this is written with "tight" domain at the schema level, e.g. if it's 100x200 sparse the domain will be `((0,99),(0,199))`
  * You can look at non-empty domain at the data level but that might be say `((2,99),(0,198))` or somesuch if an entire row or column is entirely unoccupied
  * You can get its nnz using the same function above
    * However this (unlike in the fully occuped 1D case) does not_ give back the 100x200 info
    * Suppose the array is 7.4% occupied -- this will give you the single number 1,480 which is the number of IJV triples in the sparse array but nothing about the original 100 or 200
  * The issues with the previous two bullets are why we're writing "tight" domain -- so we can read back the 100 x 200 info from the domain and the domain is the place we "store" this information

TL;DR:

* For `DataFrame` which is 1D sparse but fully occupied:
  * domain, non-empty domain, or nnz all could be used to store the 'shape'
  * we use nnz, and we leave domain all-int64, and we don't use non-empty domain programatically
* For `SparseNDArray` which is 2D sparse and usually not fully occupied:
  * non-empty domain cannot be relied on to recover 'shape'
  * nor can nnz
  * so we use 'tight' domain to store the 'shape'

## Options

* If we continue to use domain to store the 100 x 200 information:
  * Add a 'write with tight domain or not' flag as Bruce suggests
    * Anyone wanting 'tight' domain:
      * Their `X` matrices are non-growable
      * They can get 'shape' from domain
    * Anyone wanting 'growable' domain:
      * Their `X` matrices are growable
      * They can't get 'shape' from domain
        * Maybe they live without that
        * Maybe we store  the 100 x 200 info as array metadata
* If we store the 100 x 200 information somewhere other than the domain:
  * Maybe as array metadata?
  * Then the API wouldn't need a 'write with tight domain or not' flag since we would no longer be using the domain to store the shape
  * `X` matrices will be growable
  * 'shape' will be gettable from non-domain storage

See also #961 for a related but distinct issue which is explicit shaping of `DataFrame`.

**Comment by johnkerl** [comment]
Proposal from @Shelnutt2 and @johnkerl: 

* Spec: at `SOMADataFrame.create`
  * [create](https://github.com/single-cell-data/SOMA/blob/python-v0.0.0a13/python-spec/src/somacore/data.py#L183-L191)
  * Add a new optional kwarg `immutable=False`
  * Disallow `immutable=True` for `DenseNDArray`
* Python impl:
  * [create](https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/apis/python/src/tiledbsoma/common_nd_array.py#L23-L32)
  * Add a new optional kwarg `immutable=False`
  * Disallow `immutable=True` for `DenseNDArray`
  * [shape](https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/apis/python/src/tiledbsoma/common_nd_array.py#L61-L66)
  * If mutable (detected via `shape` slot `[0]` being 0 and slot `[1]` being maxint64) then return the non-empty domain
  * Update docstrings to indicate that if say `n_obs=100` and `n_var=200` and in the case where the first or last row or column of the `SparseNDArray` is entirely unpopulated, `shape` will not be `(100,200)` but will rather reflect the non-empty domain
  * Point out to the user that they can also consult the experiment/measurement's `obs.count` and `var.count`
* This will allow `soma.io` users to expand `X` arrays which is a feature they currently have in `main-old` but do not in `main` (a regression)

@bkmartinjr thoughts?

**Comment by bkmartinjr** [comment]
@johnkerl - apologies, I'm having a hard time parsing this proposal as written.  I'm also concerned that it is potentially a breaking change to the current API.  Which is to say, perhaps we should convene an in-person discussion soon, prior to 1.0 RC?

**Comment by johnkerl** [comment]
Alternative suggestion (as meeting prep @bkmartinjr @Shelnutt2 ):

* Have the `shape` arg to `create` be optional-none
  * If it is `None` then it will default to 2**63-1 on each dim
  * I do not want any to have to keystroke out `2**63-1` hence my proposal that they will be able to say `None`
  * This is a non-breaking change to the API
* For the `shape` method:
 * If the domain on a dim goes to 2**63-1 then return non-empty domain
 * Else return domain
 * This is a non-breaking change to the API

**Comment by johnkerl** [comment]
Updated after discussion between @bkmartinjr @Shelnutt2 @johnkerl :

* For `SOMADataFrame` and `SparseNDArray` (the two sparse foundational types):
  * The `shape` argument to `create` is to be interpreted as how big the array _could be_ -- its capacity
  * The `.shape` accessor is to be interpreted as how big the array _could be_ -- its capacity -- this simply returns what the `shape` argument to `create` was when the array was created
  * Docstrings will be updated to indicate this
  * The `.count` accessor for `SOMADataFrame` (which exists) returns how big the dataframe _currently is_
  * For `SparseNDArray` (for the reasons detailed above) -- we will tell the users that if they want to see how big it _actually is_ then they should consult `obs.count` and `var.count`
* For `SOMADataFrame`:
  * There is currently no `shape` argument to `create` but their should be, and it should behave as above 
* For `soma.io`:
  * It needs to pass in `shape` slots being 2**63-1 slotwise (or, `None` slotwise as aliases for 2**63-1)
* For `DenseNDArray`:
  * The current size and the capacity are the same. No changes will be allowed here.

**Comment by johnkerl** [comment]
A complication at the implementation level: these will need to be dealt with as well:

* https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/apis/python/src/tiledbsoma/read_iters.py#L73
* https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/apis/python/src/tiledbsoma/read_iters.py#L94

These depend on the current semantics of shape which are _how big the array currently is_ (the status quo), not _how big it could be_ (what is proposed on this issue).

**Comment by johnkerl** [comment]
Discussion with @bkmartinjr and @thetorpedodog: 

* In our old "immutable X" world, `.shape` was both the _current bounds_ as well as the _future capacity (the same)
* In our new "mutable X" world, `.shape` is only the latter
* `.csrs`, `.cscs`, and `.coos` rely on the former semantics
* However they are _only_ convenience functions -- and moreover they are not being used
  * https://github.com/single-cell-data/TileDB-SOMA/blob/0.5.0a11/apis/python/src/tiledbsoma/read_iters.py#L65-L114
* We will jettison these at both the spec and impl levels -- at this which point the rest of the work on this issue is trivial
  * Validated on WIP branches -- PRs coming soon

**Comment by pablo-gar** [comment]
Since the priority tag has been used I wanted to ask if this required for 1.0 RC?

**Comment by bkmartinjr** [comment]
> Since the priority tag has been used I wanted to ask if this required for 1.0 RC?

Yes.  Hard yes.

**Comment by johnkerl** [comment]
@pablo-gar this is a @Shelnutt2 and @bkmartinjr / @maniarathi call

**Comment by johnkerl** [comment]
but since @maniarathi added the `priority` tag I believe that's her way of saying "yes, required for rc0"

**Comment by bkmartinjr** [comment]
Recall that it is a breaking API change - it requires removal of some existing API.

So, given the RC definition, and importance of the feature to existing users, how can we _not_ include it?

**Comment by maniarathi** [comment]
Yes, this is a required for RC0 because as @bkmartinjr mentioned, it is a breaking API change which is the requirement to be able to cut the RC.

**Comment by johnkerl** [comment]
@maniarathi #952 is merged which is the rc blocker.

R work needs to be done here -- hence my not closing this issue entirely -- but the R work is non-rc-blocking.

**Comment by johnkerl** [comment]
Splitting out an R task for tracking
